federationexample/README.md

```diff
@@ -0,0 +1,135 @@
+```
+type Device struct {
+    Id            int64
+    OrgId         int64
+    Name          string
+}
+device := schema.Object("Device", Device{})
+```
+
+A federated object has an extra field called FetchObjectFromKeys that takes a function, telling the gqlserver how to fecth the full object from a the federated keys, the minimum number of fields needed to fetch the object. A federated object can exist on multiple gqlservers and hhave different field funcs on each of the team's gqlservers. 
+
+```
+type Device struct {
+    Id            int64
+    OrgId         int64
+    Name          string
+}
+// Device keys has to be a subset of the fields on Device and is the minimum information needded to fetch a full Device
+type DeviceKeys struct {
+    Id            int64
+    OrgId         int64
+}
+device := schema.Object("Device", Device{}, schemabuilder.FetchObjectFromKeys(func(args struct{ Keys []*DeviceKeys }) []*Device {
+    // Fetch device from database using device keys    
+    return fetchedDevices
+}))
+```
+
+
+## Example
+
+In teh federation example, we have 2 gqlservers (device and safety). Device and Location are federated objects, whcih allows them to have field funcs on both gqlservers. 
+```
+{
+  device {                  #devicegqlserver
+    id                      #devicegqlserver
+    location {              #devicegqlserver
+      latitude              #devicegqlserver
+      longitude             #devicegqlserver
+      street                #devicegqlserver
+    }
+    camera{                 #safetygqlserver
+      id                    #safetygqlserver
+      isOn                  #safetygqlserver
+      isMultiCam            #safetygqlserver
+      cameraLocation {      #safetygqlserver
+      	latitude            #safetygqlserver
+      	longitude           #safetygqlserver
+        street              #devicegqlserver
+   	  }
+    }
+  }
+}
+```
+
+
+When the gateway parses this query, it first makes a request to devicegqlserver. It adds all the fields that exist on device server, and when it sees a field it doesnt know how to resolve it replalces it with the field func `__federation` and adds theh deviceKeys are fields nested on `__federation`.
+```
+{
+  device {                  #devicegqlserver
+    id                      #devicegqlserver
+    location {              #devicegqlserver
+      latitude              #devicegqlserver
+      longitude             #devicegqlserver
+      street                #devicegqlserver
+    }
+    _federation {
+        id
+        orgId
+    }
+  }
+}
+
+```
+
+Once it gets the repsonse, it makes another request to safetygqlserver to resolve the camera field. The args `deviceKeys` are a list of he `DeviceKey` struct with the id and org Id populated from theh results of the previous query. The field `Device-safetygqserver` runs the function passed into `FetchObjectFromKeys` in the object registration, telling it how to get the full device from the keys. 
```

Comment:
```suggestion
Once the gateway receives the response from devicegqlserver, it forwards another sub-query to safety-gqlserver to resolve the camera fields. The args `deviceKeys` are a list of `DeviceKey` struct with `id` and `orgId` populated from the results of the previous sub-query to devicegqlserver. The field `Device_safetygqserver` runs the function passed into `FetchObjectFromKeys` in the object registration, telling it how to get the full device from the keys. 
```

Reply:
> The field `Device_safetygqserver` runs the function passed into `FetchObjectFromKeys` in the object registration, telling it how to get the full device from the keys. 

I'm a bit confused about this statement.

Reply:
The field `Device_safetygqserver` executes the `FetchObjectFromKeys` function defined on the device object. It takes a list of `deviceKeys` and returns a list of  `Device` objects.